### PR TITLE
[SIMS-#338]- Part-Time Ecert Feedback file

### PR DESCRIPTION
**Pull request by guru-aot**
- [x] Using a similar process to CRA, create fixed file to READ file from zoneB SFTP server
- [x] Update db values from feedback file 
- [x] Schedule it in camunda

Additional taks done
- [x] Refactorin fulltime and parttime Ecert feedback file by adding abstract implementations
- [x] Added the missed NOA forms json which was a change from previous PR ECert request processing
- [x] Renamed existing scheduler to append Part-Time and Full-Time in the names

### CODE REVIEW

File: sources/packages/forms/src/form-definitions/noticeofassessment.json [hunk 1]
```diff
@@ -13305,7 +13504,7 @@
                         "value": ""
                       }
                     ],
-                    "content": "${{((data.disbursement?.disbursement1cspt ? parseFloat(data.disbursement.disbursement1cspt) : 0) + (data.disbursement?.disbursement2cspt ? parseFloat(data.disbursement.disbursement2cspt) : 0)).toFixed(2)}}",
```
Comment (guru-aot) [inline sources/packages/forms/src/form-definitions/noticeofassessment.json:13308]: This is the only change I did as part of the Ecert request file creation, other changes are either id or addons tags generated or people changed the NOA and did not check in to GITHUB

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import { RecordTypeCodes } from "../models/e-cert-integration-model";
+
+/**
+ * Footer of an E-Cert file.
+ * The documentation about it is available on the document
+ * 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+ */
+@Injectable()
+export abstract class ECertFileFooter implements FixedFormatFileLine {
+  recordTypeCode: RecordTypeCodes;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts:12]: I can see that these 3 variables are declared in the base class and the inherited classes, can they be declared only at the base class?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  totalSINHash: number;
+  recordCount: number;
+  public getFixedFormat(): string {
+    throw new Error("Method not implemented.");
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-footer.ts:16]: If there are no implementations I would prefer that the class is declared as `abstract`. Is there anything preventing it?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-header.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Header of an E-Cert file.
+ * The documentation about it is available on the document
+ * 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+ */
+@Injectable()
+export abstract class ECertFileHeader implements FixedFormatFileLine {
+  recordTypeCode: RecordTypeCodes;
+  public getFixedFormat(): string {
+    throw new Error("Method not implemented.");
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-header.ts:14]: If there are no implementations the method can just be abstract also, right?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-header.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import { RecordTypeCodes } from "../models/e-cert-integration-model";
+
+/**
+ * Header of an E-Cert file.
+ * The documentation about it is available on the document
+ * 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+ */
+@Injectable()
+export abstract class ECertFileHeader implements FixedFormatFileLine {
+  recordTypeCode: RecordTypeCodes;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-header.ts:12]: It seems that the base class declares 1 common property that is also declared in all derived classes. Can it be declared only on the base class assuming that they will be mandatories?
Also, there are other 2 variables declared on both derived classes, should we move them to the base?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-record.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * Record of an Entitlement E-Cert file.
+ * The documentation about it is available on the document
+ * 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+ */
+@Injectable()
+export abstract class ECertFileRecord implements FixedFormatFileLine {
+  public getFixedFormat(): string {
+    throw new Error(`Method not implemented.`);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-record.ts:12]: Please convert to an abstract method.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-record.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import { FixedFormatFileLine } from "../../../services/ssh/sftp-integration-base.models";
+
+/**
+ * Record of an Entitlement E-Cert file.
+ * The documentation about it is available on the document
+ * 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+ */
+@Injectable()
+export abstract class ECertFileRecord implements FixedFormatFileLine {
+  public getFixedFormat(): string {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-file-record.ts:11]: We are not using "explicit" public, when possible, please remove it.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+export abstract class ECertResponseRecord extends ECertResponseRecordIdentification {
+  constructor(line: string, lineNumber: number) {
+    super(line, lineNumber);
+  }
+
+  /**
+   * Financial document number associated with this disbursement.
+   */
+  public get documentNumber(): number {
+    throw new Error(`Method not implemented.`);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:20]: Please convert the methods without implementations to abstract methods.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  /**
+   * error code-1 associated with the document number.
+   * Index of the first error code is 731 not 641
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:25]: I did not follow this comment at this level. Is it supposed to be here at the base class or in the class implementing it?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * Financial document number associated with this disbursement.
+   */
+  public get documentNumber(): number {
+    throw new Error(`Method not implemented.`);
+  }
+
+  /**
+   * error code-1 associated with the document number.
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:24]: Start the sentences with upper case, please. Same for other places in this file.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts [hunk 11]
```diff
@@ -1,31 +1,121 @@
...
    * @param ecertRecords
    * @param fileSequence
    */
   createRequestContent(
     ecertRecords: ECertRecord[],
-    fileSequence: number,
+    fileSequence?: number,
   ): FixedFormatFileLine[] {
     throw new Error(
       `Method not implemented , ${ecertRecords} && ${fileSequence} not declared.`,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts:29]: This method seems to be mandatory to every class deriving from `ECertIntegrationService` can we please make it abstract?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * Financial document number associated with this disbursement.
+   */
+  public get documentNumber(): number {
+    return parseInt(this.line.substring(467, 474));
+  }
+
+  /**
+   * error code-1 associated with the document number.
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts:24]: Please make all sentences with a capital letter at the start.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+ */
+export class ECertPartTimeResponseRecord extends ECertResponseRecord {
+  constructor(line: string, lineNumber: number) {
+    super(line, lineNumber);
+  }
+
+  /**
+   * Financial document number associated with this disbursement.
+   */
+  public get documentNumber(): number {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts:19]: Please avoid using the 'public'.

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts [hunk 14]
```diff
@@ -1,5 +1,12 @@
-import { DisbursementValue, RelationshipStatus } from "../../database/entities";
+import {
+  DisbursementValue,
+  RelationshipStatus,
+} from "../../../database/entities";
 
+export const ECERT_SENT_TITLE = "ENTITLEMENT";
+export const CSGD = "CSGD";
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:7]: Please add some comments explaining what is each one, if we have this information.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts (hunk 10 above)
Comment (ann-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:24]: caps E

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+   * Index of the first error code is 731 not 641
+   * because `Grant Award Code` and `Grant Award Amount`
```
Comment (ann-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:26]: 👍

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts [hunk 16]
```diff
@@ -36,3 +43,27 @@ export type Award = Pick<
   DisbursementValue,
   "valueType" | "valueCode" | "valueAmount"
 >;
+
+/**
+ * Codes used to start all the lines of the e-Cert
+ * files sent to ESDC.
+ */
+export enum RecordTypeCodes {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:51]: Is it safe to share the same enum between PartTime and FullTime?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+   * be will repeating 10 times, which is not mentioned
+   * in the doc 'CSLP-AppendixF2AsReviewed2016-FileLayouts BC Files V3(HAJ-CB EDITS) In ESDC Folder'.
+   */
+  public get errorCode1(): string {
+    throw new Error(`Method not implemented.`);
```
Comment (ann-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:31]: why errors are thrown in all functions?

File: sources/packages/api/src/route-controllers/esdc-integration/e-cert-integration.system.controller.ts [hunk 18]
```diff
@@ -38,6 +33,22 @@ export class ECertIntegrationController extends BaseController {
...
 
+  /**
+   * Download all files from FullTime E-Cert Response folder on SFTP and process them all.
+   * @returns Summary with what was processed and the list of all errors, if any.
+   */
+  @Post("process-full-time-responses")
+  async processFullTimeResponses(): Promise<ESDCFileResponseDTO[]> {
+    const fullTimeResults =
+      await this.eCertFileHandler.processFullTimeResponses();
+    return fullTimeResults.map((fullTimeResult) => {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/esdc-integration/e-cert-integration.system.controller.ts:44]: The code can be simplified.
```typescript
return fullTimeResults.map((fullTimeResult) => ({
      processSummary: fullTimeResult.processSummary,
      errorsSummary: fullTimeResult.errorsSummary,
}));
```

File: sources/packages/api/src/route-controllers/esdc-integration/e-cert-integration.system.controller.ts [hunk 19]
```diff
@@ -57,16 +68,17 @@ export class ECertIntegrationController extends BaseController {
...
    */
-  @Post("process-responses")
-  async processResponses(): Promise<ESDCFileResponseDTO[]> {
-    const results = await this.eCertFullTimeResponseService.processResponses();
-    return results.map((result) => {
+  @Post("process-part-time-responses")
+  async processPartTimeResponses(): Promise<ESDCFileResponseDTO[]> {
+    const partTimeResults =
+      await this.eCertFileHandler.processPartTimeResponses();
+    return partTimeResults.map((partTimeResult) => {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/esdc-integration/e-cert-integration.system.controller.ts:78]: The return can be simplified.

File: sources/packages/api/src/utilities/system-configurations-constants.ts [hunk 20]
```diff
@@ -59,3 +59,10 @@ export const ECERT_PART_TIME_FILE_CODE = "PBC.EDU.PTCERTS.D";
  */
 export const MSFAA_FULL_TIME_FILE_CODE = "PBC.EDU.MSFA.SENT.";
 export const MSFAA_PART_TIME_FILE_CODE = "PBC.EDU.MSFA.SENT.PT.";
+
+/**
+ * These constants are used to specify the filename code
+ * created for Full-Time/ Part-Time feedback file.
+ */
+export const ECERT_FULL_TIME_FEEDBACK_FILE_CODE = "EDU.PBC.CERTSFB.*\\.DAT";
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/utilities/system-configurations-constants.ts:67]: `FILE_CODE` does not seems accurate. It seems that this should be something like `FILE_NAME`. I can see other constants in the same file following the same, so please rename all or leave them as it is.

File: sources/packages/api/src/utilities/disbursements-utils.ts [hunk 21]
```diff
@@ -1,4 +1,4 @@
-import { Award } from "src/esdc-integration/e-cert-integration/e-cert-integration-model";
+import { Award } from "src/esdc-integration/e-cert-integration/models/e-cert-integration-model";
```
Comment (ann-aot) [inline sources/packages/api/src/utilities/disbursements-utils.ts:1]: please remove "src/"

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, overall looks good, and thanks for reducing the duplication 👍 
Please take a look at the comments.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts [hunk 22]
```diff
@@ -1,31 +1,121 @@
...
+> {
   /**
    * This method will be implemented in the extended class and is used to create the ECert request content.
    * @param ecertRecords
    * @param fileSequence
    */
   createRequestContent(
     ecertRecords: ECertRecord[],
-    fileSequence: number,
+    fileSequence?: number,
```
Comment (ann-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts:26]: why is this optional?

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts (hunk 17 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:31]: They are supposed to be implemented in the derived classes @ann-aot, but I asked @guru-aot to convert those to abstract methods.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
+import { ECertResponseRecord } from "../../e-cert-files/e-cert-response-record";
+
+/**
+ * Parsing Part-Time E-Certs feedback records to get the document number and associated
+ * error codes
```
Comment (ann-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-part-time-integration/e-cert-files/e-cert-response-record.ts:5]: full stop

**Review by ann-aot — COMMENTED** [review]
Good Work @guru-aot added some comments

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts (hunk 22 above)
Comment (guru-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-integration.service.ts:26]: For fulltime implementation, the filesequence is used in the record file, but for parttime its not used.

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts (hunk 9 above)
Comment (guru-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts:25]: Copy paste issue :wink

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts (hunk 16 above)
Comment (guru-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:51]: I think its fine, as long as we have the variables declared in it are different. I did this as the declaration variables in the base class can be reused.

File: sources/packages/api/src/utilities/system-configurations-constants.ts (hunk 20 above)
Comment (guru-aot) [reply sources/packages/api/src/utilities/system-configurations-constants.ts:67]: I thought of it more like a filename Code, if you see the comments, I have explained. Eventually as this is a variable used in the regex of the filename, i thought to go with code. Didnt think a lot on the variable name, it came eventually

File: sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts (hunk 10 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/e-cert-files/e-cert-response-record.ts]: Please search for `error code-` in the source code. There are still some occurrences.

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts (hunk 14 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts]: Please ensure that the comments are added as below to allow the IntelliSense to work.
```typescript
/**
 * Amount of Grant for Part-time Studies (CSGP-PT) at the study start.
 */
export const CSGD = "CSGD";
```

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts (hunk 16 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:60]: I still think that these are areas that we talked about not "merging" and keeping them segregated would be better for maintenance in the future. Open for discussion.

**Review by andrewsignori-aot — APPROVED** [review]
Great work, looks good 👍

**Review by ann-aot — APPROVED** [review]
👍 Good Work @guru-aot . Thanks for doing the changes

**Pull request merged**